distribution/cachedir/manager.go

```
@@ -0,0 +1,214 @@
+	// still considered released.
+	DeleteDir(key string) error
+	// CollectGarbage deletes all dirs that are not currently in use (ones that
+	// don't have a GetDir without a matching ReleaseDir/DeleteDir call) and
+	// returns the amount of space saved in bytes.
+	CollectGarbage() (uint64, error)
+	// Usage returns information about how much space the download cache is
+	// using. If any directories or files are unreadable, we just log the
+	// errors and don't count the files.
+	Usage() types.DownloadCacheUsage
+}
+
+type manager struct {
+	cacheRoot string
+	mu        sync.Mutex
+	inUse     map[digest.Digest]struct{}
+}
+
+// NewManager creates a new cachedir manager
+func NewManager(path string) Manager {
+	return &manager{
+		cacheRoot: path,
+		inUse:     map[digest.Digest]struct{}{},
+	}
+}
+
+func (m *manager) cachePathForKeyHash(keyHash digest.Digest) string {
+	return filepath.Join(m.cacheRoot, keyHash.Algorithm().String(), keyHash.Hex())
+}
+
+func dirSize(path string) (uint64, error) {
+	var size uint64
+	err := filepath.Walk(path, func(_ string, info os.FileInfo, err error) error {
+		if info != nil && !info.IsDir() {
+			size += uint64(info.Size())
+		}
+		return err
+	})
+	return size, err
+}
+
+func (m *manager) sizeEstimateByKeyHash(keyHash digest.Digest) uint64 {
+	dir := m.cachePathForKeyHash(keyHash)
+	size, err := dirSize(dir)
+	if err != nil {
+		logrus.WithError(err).WithField("cachedir", dir).Warn("failed to determine size of directory")
+		return 0
+	}
+	return size
+}
+
+func (m *manager) cleanUpByKeyHash(keyHash digest.Digest) error {
+	return os.RemoveAll(m.cachePathForKeyHash(keyHash))
+}
+
+func (m *manager) GetDir(key string) string {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	hashedKey := digest.FromString(key)
+	logrus.WithField("key", key).WithField("hashedKey", hashedKey).Debug("cachedir manager get")
+	if _, ok := m.inUse[hashedKey]; ok {
+		panic(fmt.Errorf("tried to get a cachedir key (%s) twice without releasing/deleting", key))
```

I'd return an error here, and have the download manager panic if this function returns an error. Subtle difference, but it means this function doesn't have a hidden precondition.